[JSC] Add fast path for iteration on Map

sosukesuzuki opened the pull request:
#### 38df8fb7a13a27c7305fb927e6d7094b12f22ef2
<pre>
[JSC] Add fast path for iteration on Map
<a href="https://bugs.webkit.org/show_bug.cgi?id=298171">https://bugs.webkit.org/show_bug.cgi?id=298171</a>

Reviewed by Keith Miller.

This patch changes to apply the same optimization as commits.webkit.org/299358@main
to Map.

                                        TipOfTree                  Patched

iterator-prototype-forEach-map        4.6865+-0.1369     ^      0.9110+-0.0499        ^ definitely 5.1442x faster

* JSTests/microbenchmarks/iterator-prototype-forEach-map.js: Added.
(test):
* JSTests/microbenchmarks/iterator-prototype-toArray-map.js: Added.
(test):
* JSTests/stress/aggregate-error-constructor-broken-map.js: Added.
(shouldBe):
(shouldThrow):
(throw.new.Error):
(shouldBe.map.Symbol.iterator):
(mapIterator.next):
(new.AggregateError):
* JSTests/stress/iterator-prototype-forEach-map.js: Added.
(test):
(shouldBe):
(throw.new.Error):
(shouldBe.Map.prototype.Symbol.iterator):
* JSTests/stress/iterator-prototype-toArray-broken-map-iterator.js: Added.
(test):
(shouldBe):
(shouldThrow):
(throw.new.Error.mapIterator.next):
(throw.new.Error):
(newProto.next):
* JSTests/stress/iterator-prototype-toArray-map.js: Added.
(test):
(shouldBe):
(throw.new.Error):
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/runtime/IteratorOperations.h:
(JSC::forEachInMapStorage):
(JSC::forEachInIterable):
(JSC::forEachInIteratorProtocol):
* Source/JavaScriptCore/runtime/JSMap.cpp:
(JSC::JSMap::isIteratorProtocolFastAndNonObservable): Deleted.
* Source/JavaScriptCore/runtime/JSMap.h:
* Source/JavaScriptCore/runtime/JSMapInlines.h:
(JSC::JSMap::isIteratorProtocolFastAndNonObservable):
* Source/JavaScriptCore/runtime/MapIteratorPrototype.h:
* Source/JavaScriptCore/runtime/MapIteratorPrototypeInlines.h:
(JSC::mapIteratorProtocolIsFastAndNonObservable):

Canonical link: <a href="https://commits.webkit.org/299751@main">https://commits.webkit.org/299751@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9ce4fa5f0e61f4d98793c10992e4af13265aa5ac

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/120008 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39701 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/30352 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/126345 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/72081 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40397 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/48278 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91133 "") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60448 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122960 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32267 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107609 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71688 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31300 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25713 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69977 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/112133 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101744 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25901 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/129258 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/118524 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46928 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35587 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99751 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/47294 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103796 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99596 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25300 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45054 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23072 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/43549 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46790 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/52496 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/147223 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/46256 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37831 "Found 1 new JSC binary failure: testapi, Found 18665 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Array/array_slice.js.default, ChakraCore.yaml/ChakraCore/test/Array/push2.js.default, ChakraCore.yaml/ChakraCore/test/Array/reverse1.js.default, ChakraCore.yaml/ChakraCore/test/Array/toLocaleString.js.default, ChakraCore.yaml/ChakraCore/test/Bugs/blue_1086262.js.default, ChakraCore.yaml/ChakraCore/test/Bugs/bug215238.shrua-2.js.default, ChakraCore.yaml/ChakraCore/test/Date/date_cache_bug.js.default, ChakraCore.yaml/ChakraCore/test/Error/CallNonFunction.js.default, ChakraCore.yaml/ChakraCore/test/Error/stack2.js.default, ChakraCore.yaml/ChakraCore/test/Function/FuncBody.bug227901.js.default ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49605 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47942 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/620b870676f5235cb3751e20f607fb2132cca0af. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50121

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/0776670cef819ca31762db4bfef97b8e634be04f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50121

Jarred-Sumner commented on Source/JavaScriptCore/runtime/IteratorOperations.h:
```diff
@@ -63,6 +67,43 @@ JS_EXPORT_PRIVATE IterationMode getIterationMode(VM&, JSGlobalObject*, JSValue i
+
+        storage = jsCast<JSMap::Storage*>(storageCell);
+        entry = JSMap::Helper::iterationEntry(*storage) + 1;
+        JSValue entryKey = JSMap::Helper::getIterationEntryKey(*storage);
```
maybe it should only get the specific value if it would be used? like if keys || entries, then entryKey, if values || entries, then entryValue?

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b337c55ceda23217c7b5d270e6e95cc5af999fd1. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50121

Constellation commented on Source/JavaScriptCore/runtime/JSMap.h:
```diff
@@ -59,7 +59,7 @@ class JSMap final : public OrderedHashMap {
 
     static bool isSetFastAndNonObservable(Structure*);
-    bool isIteratorProtocolFastAndNonObservable();
+    JS_EXPORT_PRIVATE bool isIteratorProtocolFastAndNonObservable();
```
This is concerning since this prohibits inlining this check.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/97c5ed124a386a174d6549bc7991ea3be7bc2334. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50121

kmiller68 commented on Source/JavaScriptCore/runtime/IteratorOperations.h:
```diff
@@ -214,7 +265,22 @@ void forEachInIteratorProtocol(JSGlobalObject* globalObject, JSValue iterable, N
 
-    if (auto* setIterator = jsDynamicCast<JSSetIterator*>(iterable)) {
+    if (auto* mapIterator = jsDynamicCast<JSMapIterator*>(iterable)) {
+        if (mapIteratorProtocolIsFastAndNonObservable(vm, mapIterator)) {
+            if (JSMap* iteratedMap = jsCast<JSMap*>(mapIterator->iteratedObject())) {
+                if (iteratedMap->isIteratorProtocolFastAndNonObservable()) {
```
Why do you need both of these checks? Once the iterator says it's not observable you shouldn't need to check the map again? I suppose ditto for the Set case below too.

sosukesuzuki replied:
You're right! I removed the second check

sosukesuzuki commented on JSTests/microbenchmarks/iterator-prototype-toArray-map.js:
```diff
@@ -0,0 +1,13 @@
+}
+
+let r;
+for (let i = 0; i < 1e4; i++) {
```
Fixed

sosukesuzuki commented on Source/JavaScriptCore/runtime/IteratorOperations.h:
```diff
@@ -63,6 +67,44 @@ JS_EXPORT_PRIVATE IterationMode getIterationMode(VM&, JSGlobalObject*, JSValue i
+        case IterationKind::Entries: {
+            JSValue entryKey = JSMap::Helper::getIterationEntryKey(*storage);
+            JSValue entryValue = JSMap::Helper::getIterationEntryValue(*storage);
+            value = createTuple(globalObject, entryKey, entryValue);
```
Fixed

kmiller68 reviewed (approved):
r=me

sosukesuzuki committed:
[JSC] Add fast path for iteration on Map
https://bugs.webkit.org/show_bug.cgi?id=298171

Reviewed by Keith Miller.

This patch changes to apply the same optimization as commits.webkit.org/299358@main
to Map.

                                        TipOfTree                  Patched

iterator-prototype-forEach-map        4.6865+-0.1369     ^      0.9110+-0.0499        ^ definitely 5.1442x faster

* JSTests/microbenchmarks/iterator-prototype-forEach-map.js: Added.
(test):
* JSTests/microbenchmarks/iterator-prototype-toArray-map.js: Added.
(test):
* JSTests/stress/aggregate-error-constructor-broken-map.js: Added.
(shouldBe):
(shouldThrow):
(throw.new.Error):
(shouldBe.map.Symbol.iterator):
(mapIterator.next):
(new.AggregateError):
* JSTests/stress/iterator-prototype-forEach-map.js: Added.
(test):
(shouldBe):
(throw.new.Error):
(shouldBe.Map.prototype.Symbol.iterator):
* JSTests/stress/iterator-prototype-toArray-broken-map-iterator.js: Added.
(test):
(shouldBe):
(shouldThrow):
(throw.new.Error.mapIterator.next):
(throw.new.Error):
(newProto.next):
* JSTests/stress/iterator-prototype-toArray-map.js: Added.
(test):
(shouldBe):
(throw.new.Error):
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/runtime/IteratorOperations.h:
(JSC::forEachInMapStorage):
(JSC::forEachInIterable):
(JSC::forEachInIteratorProtocol):
* Source/JavaScriptCore/runtime/JSMap.cpp:
(JSC::JSMap::isIteratorProtocolFastAndNonObservable): Deleted.
* Source/JavaScriptCore/runtime/JSMap.h:
* Source/JavaScriptCore/runtime/JSMapInlines.h:
(JSC::JSMap::isIteratorProtocolFastAndNonObservable):
* Source/JavaScriptCore/runtime/MapIteratorPrototype.h:
* Source/JavaScriptCore/runtime/MapIteratorPrototypeInlines.h:
(JSC::mapIteratorProtocolIsFastAndNonObservable):

Canonical link: https://commits.webkit.org/299751@main
